File: app_manager/appmanager/src/main/java/app/editors/manager/managers/utils/FirebaseUtils.java
package app.editors.manager.managers.utils;

import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.google.firebase.analytics.FirebaseAnalytics;
import com.google.firebase.crashlytics.FirebaseCrashlytics;
import com.google.firebase.remoteconfig.FirebaseRemoteConfig;
import com.google.firebase.remoteconfig.FirebaseRemoteConfigSettings;

import app.editors.manager.BuildConfig;
import app.editors.manager.R;
import app.editors.manager.app.App;

public class FirebaseUtils {

    public static final class AnalyticsEvents {
        public static final String CREATE_PORTAL = "portal_create";
        public static final String CREATE_ENTITY = "create_entity";
        public static final String CHECK_PORTAL = "check_portal";
        public static final String LOGIN_PORTAL = "portal_login";
        public static final String SWITCH_ACCOUNT = "account_switch";
        public static final String OPEN_PDF = "open_pdf";
        public static final String OPEN_EDITOR = "open_editor";
        public static final String OPEN_MEDIA = "open_media";
        public static final String OPEN_EXTERNAL = "open_external";
        public static final String OPERATION_RESULT = "operation_result";
        public static final String OPERATION_DETAILS = "operation_details";
    }

    public static final class AnalyticsKeys {
        public static final String NONE = "none";
        public static final String SUCCESS = "success";
        public static final String FAILED = "failed";
        public static final String PORTAL = "portal";
        public static final String LOGIN = "email";
        public static final String PROVIDER = "provider";
        public static final String ON_DEVICE = "onDevice";
        public static final String TYPE = "type";
        public static final String FILE_EXT = "fileExt";
        public static final String TYPE_FILE = "file";
        public static final String TYPE_FOLDER = "folder";
    }

    private static final String KEY_BETA = "android_documents_20_beta_enabled";
    private static final String KEY_VERSION = "android_documents_20_beta_version";
    private static final String KEY_RATING = "android_documents_rating";

    private static final long TIME_FETCH = 3600;

    private static FirebaseRemoteConfig sRemoteConfig = null;
    private static FirebaseAnalytics sFirebaseAnalytics = null;

//    public interface OnBeta {
//        void onBetaVersion(boolean isBeta, long version);
//    }

    public interface OnRatingApp {
        void onRatingApp(boolean isRating);
    }

    public static void addCrash(@NonNull final String message) {
        if (App.getApp().isAnalyticEnable()) {
            FirebaseCrashlytics.getInstance().log(message);
        }
    }

    public static void addCrash(@NonNull final Throwable throwable) {
        if (App.getApp().isAnalyticEnable()) {
            FirebaseCrashlytics.getInstance().recordException(throwable);
        }
    }

    private static void initRemoteConfig() {
        if (sRemoteConfig == null && App.getApp().isAnalyticEnable()) {
            sRemoteConfig = FirebaseRemoteConfig.getInstance();
            final FirebaseRemoteConfigSettings configSettings = new FirebaseRemoteConfigSettings.Builder()
                    .setFetchTimeoutInSeconds(3600L)
                    .build();
            sRemoteConfig.setConfigSettingsAsync(configSettings);
            sRemoteConfig.setDefaultsAsync(R.xml.remote_config_defaults);
        }
    }

//    //TODO Не испрользовать бета версию
//    public static void checkBetaConfig(@Nullable final OnBeta onBeta) {
//        initRemoteConfig();
//        sRemoteConfig.fetch(BuildConfig.DEBUG ? 0 : TIME_FETCH).addOnCompleteListener(task -> {
//            if (task.isSuccessful()) {
//                sRemoteConfig.activate();
//                final boolean isBeta = sRemoteConfig.getBoolean(KEY_BETA);
//                final long version = sRemoteConfig.getLong(KEY_VERSION);
//                if (onBeta != null) {
//                    onBeta.onBetaVersion(isBeta, version);
//                }
//            }
//        });
//    }

    public static void checkRatingConfig(@Nullable final OnRatingApp onRatingApp) {
        if (App.getApp().isAnalyticEnable()) {
            initRemoteConfig();
            sRemoteConfig.fetch(BuildConfig.DEBUG ? 0 : TIME_FETCH).addOnCompleteListener(task -> {
                if (task.isSuccessful()) {
                    sRemoteConfig.activate();
                    final boolean isRatingApp = sRemoteConfig.getBoolean(KEY_RATING);
                    if (onRatingApp != null) {
                        onRatingApp.onRatingApp(isRatingApp);
                    }
                }
            });
        } else {
            sRemoteConfig = null;
        }
    }

    public static void addAnalytics(@NonNull final String event, @NonNull final Bundle bundle) {
        if (App.getApp().isAnalyticEnable()) {
            if (sFirebaseAnalytics == null) {
                sFirebaseAnalytics = FirebaseAnalytics.getInstance(App.getApp());
            }
            sFirebaseAnalytics.logEvent(event, bundle);
        } else {
            sFirebaseAnalytics = null;
        }
    }

    public static void addAnalyticsCreatePortal(@NonNull final String portal, @NonNull final String login) {
        final Bundle bundle = new Bundle();
        bundle.putString(AnalyticsKeys.PORTAL, portal);
        bundle.putString(AnalyticsKeys.LOGIN, login);
        addAnalytics(AnalyticsEvents.CREATE_PORTAL, bundle);
    }

    public static void addAnalyticsCheckPortal(@NonNull final String portal, @NonNull final String result, @Nullable final String error) {
        final Bundle bundle = new Bundle();
        bundle.putString(AnalyticsKeys.PORTAL, portal);
        bundle.putString(AnalyticsEvents.OPERATION_RESULT, result);
        bundle.putString(AnalyticsEvents.OPERATION_DETAILS, error != null ? error : AnalyticsKeys.NONE);
        addAnalytics(AnalyticsEvents.CHECK_PORTAL, bundle);
    }

    public static void addAnalyticsLogin(@NonNull final String portal, @Nullable final String provider) {
        final Bundle bundle = new Bundle();
        bundle.putString(AnalyticsKeys.PORTAL, portal);
        bundle.putString(AnalyticsKeys.PROVIDER, provider != null ? provider : AnalyticsKeys.NONE);
        addAnalytics(AnalyticsEvents.LOGIN_PORTAL, bundle);
    }

    public static void addAnalyticsSwitchAccount(final String portal) {
        final Bundle bundle = new Bundle();
        bundle.putString(AnalyticsKeys.PORTAL, portal);
        addAnalytics(AnalyticsEvents.SWITCH_ACCOUNT, bundle);
    }

    public static void addAnalyticsCreateEntity(@NonNull final String portal, final boolean isFile, @Nullable final String extension) {
        final Bundle bundle = new Bundle();
        bundle.putString(AnalyticsKeys.PORTAL, portal);
        bundle.putString(AnalyticsKeys.ON_DEVICE, "false");
        bundle.putString(AnalyticsKeys.TYPE, isFile ? AnalyticsKeys.TYPE_FILE : AnalyticsKeys.TYPE_FOLDER);
        bundle.putString(AnalyticsKeys.FILE_EXT, extension != null ? extension : AnalyticsKeys.NONE);
        addAnalytics(AnalyticsEvents.CREATE_ENTITY, bundle);
    }

    public static void addAnalyticsOpenEntity(@NonNull final String portal, @NonNull final String extension) {
        final Bundle bundle = new Bundle();
        bundle.putString(AnalyticsKeys.PORTAL, portal);
        bundle.putString(AnalyticsKeys.ON_DEVICE, "false");
        bundle.putString(AnalyticsKeys.FILE_EXT, extension);
        addAnalytics(AnalyticsEvents.OPEN_EDITOR, bundle);
    }

    public static void addAnalyticsOpenExternal(@NonNull final String portal, @NonNull final String extension) {
        final Bundle bundle = new Bundle();
        bundle.putString(AnalyticsKeys.PORTAL, portal);
        bundle.putString(AnalyticsKeys.ON_DEVICE, "false");
        bundle.putString(AnalyticsKeys.FILE_EXT, extension);
        addAnalytics(AnalyticsEvents.OPEN_EXTERNAL, bundle);
    }

}

File: app_manager/appmanager/src/main/java/app/editors/manager/managers/utils/FirebaseUtils.kt
package app.editors.manager.managers.utils

import android.content.Context
import android.os.Bundle
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import app.documents.core.model.cloud.CloudAccount
import app.editors.manager.BuildConfig
import app.editors.manager.R
import app.editors.manager.app.App.Companion.getApp
import com.google.firebase.analytics.FirebaseAnalytics
import com.google.firebase.crashlytics.FirebaseCrashlytics
import com.google.firebase.remoteconfig.FirebaseRemoteConfig
import com.google.firebase.remoteconfig.FirebaseRemoteConfigSettings
import lib.toolkit.base.managers.utils.FileUtils
import retrofit2.HttpException

object FirebaseUtils {

    private const val KEY_RATING = "android_documents_rating"
    private const val KEY_CAPTCHA = "recaptcha_for_portal_registration"
    private const val KEY_TERMS_OF_SERVICE = "link_terms_of_service"
    private const val KEY_PRIVACY_POLICY = "link_privacy_policy"
    private const val KEY_ALLOW_COAUTHORING = "allow_coauthoring"
    private const val KEY_SDK_FULLY = "check_sdk_fully"
    private const val KEY_GOOGLE_DRIVE = "allow_google_drive"
    private const val TIME_FETCH: Long = 3600

    private var firebaseAnalytics: FirebaseAnalytics? = null

    @JvmStatic
    fun addCrash(message: String) {
        if (getApp().isAnalyticEnable) {
            FirebaseCrashlytics.getInstance().log(message)
        }
    }

    @JvmStatic
    fun addCrash(throwable: Throwable) {
        if (getApp().isAnalyticEnable) {
            FirebaseCrashlytics.getInstance().recordException(throwable)
        }
    }

    @JvmStatic
    fun addCrash(httpException: HttpException) {
        if (getApp().isAnalyticEnable) {
            val message = httpException.response()?.errorBody()?.string()
            if (message != null) addCrash(message = message) else addCrash(throwable = httpException)
        }
    }

    private fun getRemoteConfig(): FirebaseRemoteConfig? {
        return FirebaseRemoteConfig.getInstance().apply {
            setConfigSettingsAsync(
                FirebaseRemoteConfigSettings.Builder()
                    .setFetchTimeoutInSeconds(3600L)
                    .build()
            )
            setDefaultsAsync(R.xml.remote_config_defaults)
        }
    }


    fun checkRatingConfig(onRatingApp: OnRatingApp?) {
        if (getApp().isAnalyticEnable) {
            getRemoteConfig()?.let { config ->
                config.fetch(if (BuildConfig.DEBUG) 0 else TIME_FETCH).addOnCompleteListener { task ->
                    if (task.isSuccessful) {
                        config.activate()
                        val isRatingApp = config.getBoolean(KEY_RATING)
                        onRatingApp?.onRatingApp(isRatingApp)
                    }
                }
            }

        }
    }

    fun isGoogleDriveEnable(block: (isEnable: Boolean) -> Unit) {
        if (getApp().isAnalyticEnable) {
            getRemoteConfig()?.let { config ->
                config.fetch(if (BuildConfig.DEBUG) 0 else TIME_FETCH).addOnCompleteListener { task ->
                    if (task.isSuccessful) {
                        config.activate()
                        block(config.getBoolean(KEY_GOOGLE_DRIVE))
                    }
                }
            }
        }
    }

    @JvmStatic
    fun isCaptchaEnable(block: (isEnable: Boolean) -> Unit) {
        if (getApp().isAnalyticEnable) {
            getRemoteConfig()?.let { config ->
                config.fetch(if (BuildConfig.DEBUG) 0 else TIME_FETCH).addOnCompleteListener { task ->
                    if (task.isSuccessful) {
                        config.activate()
                        block(config.getBoolean(KEY_CAPTCHA))
                    }
                }
            }
        }
    }

    fun getServiceUrls(): LiveData<Array<String>?> {
        val liveData = MutableLiveData<Array<String>>(null)
        getRemoteConfig()?.let { config ->
            config.fetch(if (BuildConfig.DEBUG) 0 else TIME_FETCH).addOnCompleteListener { task ->
                if (task.isSuccessful) {
                    config.activate()
                    liveData.value =
                        arrayOf(config.getString(KEY_TERMS_OF_SERVICE), config.getString(KEY_PRIVACY_POLICY))
                }
            }
        }
        return liveData
    }

    fun checkSdkVersion(
        context: Context,
        account: CloudAccount,
        onResult: (isCoauthoring: Boolean) -> Unit
    ) {
        getSdk { allowCoauthoring, checkSdkFully ->
            if (allowCoauthoring) {
                onResult(false)
                return@getSdk
            }

            val webSdk = account
                .portal
                .version
                .documentServerVersion
                .replace(".", "")

            if (webSdk.isEmpty()) {
                onResult(false)
                return@getSdk
            }

            val localSdk = FileUtils.readSdkVersion(context).replace(".", "")

            var maxVersionIndex = 2

            if (!checkSdkFully) {
                maxVersionIndex = 1
            }

            for (i in 0..maxVersionIndex) {
                if (webSdk[i] != localSdk[i]) {
                    onResult(false)
                    return@getSdk
                }
            }

            onResult(true)
        }
    }

    /**
     * @param block First allow_coauthoring, second check_sdk_fully
     */
    fun getSdk(block: (allowCoauthoring: Boolean, checkSdkFully: Boolean) -> Unit)  {
        getRemoteConfig()?.let { config ->
            config.fetch(if (BuildConfig.DEBUG) 0 else TIME_FETCH)
                .addOnCompleteListener { task ->
                    if (task.isSuccessful) {
                        config.activate()
                        block(
                            config.getBoolean(KEY_ALLOW_COAUTHORING),
                            config.getBoolean(KEY_SDK_FULLY)
                        )
                    } else {
                        block(true, false)
                    }
                }
        } ?: {
            block(true, false)
        }
    }

    fun getLocalServicesUrl(context: Context): Array<String> {
        return arrayOf(context.getString(R.string.app_url_terms), context.getString(R.string.app_url_policy))
    }

    private fun addAnalytics(event: String, bundle: Bundle) {
        if (getApp().isAnalyticEnable) {
            if (firebaseAnalytics == null) {
                firebaseAnalytics = FirebaseAnalytics.getInstance(getApp())
            }
            firebaseAnalytics!!.logEvent(event, bundle)
        } else {
            firebaseAnalytics = null
        }
    }

    @JvmStatic
    fun addAnalyticsCreatePortal(portal: String, login: String) {
        val bundle = Bundle()
        bundle.putString(AnalyticsKeys.PORTAL, portal)
        bundle.putString(AnalyticsKeys.LOGIN, login)
        addAnalytics(AnalyticsEvents.CREATE_PORTAL, bundle)
    }

    @JvmStatic
    fun addAnalyticsCheckPortal(portal: String, result: String, error: String?) {
        val bundle = Bundle()
        bundle.putString(AnalyticsKeys.PORTAL, portal)
        bundle.putString(AnalyticsEvents.OPERATION_RESULT, result)
        bundle.putString(AnalyticsEvents.OPERATION_DETAILS, error ?: AnalyticsKeys.NONE)
        addAnalytics(AnalyticsEvents.CHECK_PORTAL, bundle)
    }

    fun addAnalyticsLogin(portal: String, provider: String?) {
        val bundle = Bundle()
        bundle.putString(AnalyticsKeys.PORTAL, portal)
        bundle.putString(AnalyticsKeys.PROVIDER, provider ?: AnalyticsKeys.NONE)
        addAnalytics(AnalyticsEvents.LOGIN_PORTAL, bundle)
    }

    @JvmStatic
    fun addAnalyticsSwitchAccount(portal: String?) {
        val bundle = Bundle()
        bundle.putString(AnalyticsKeys.PORTAL, portal)
        addAnalytics(AnalyticsEvents.SWITCH_ACCOUNT, bundle)
    }

    @JvmStatic
    fun addAnalyticsCreateEntity(portal: String, isFile: Boolean, extension: String?) {
        val bundle = Bundle()
        bundle.putString(AnalyticsKeys.PORTAL, portal)
        bundle.putString(AnalyticsKeys.ON_DEVICE, "false")
        bundle.putString(AnalyticsKeys.TYPE, if (isFile) AnalyticsKeys.TYPE_FILE else AnalyticsKeys.TYPE_FOLDER)
        bundle.putString(AnalyticsKeys.FILE_EXT, extension ?: AnalyticsKeys.NONE)
        addAnalytics(AnalyticsEvents.CREATE_ENTITY, bundle)
    }

    @JvmStatic
    fun addAnalyticsOpenEntity(portal: String, extension: String) {
        val bundle = Bundle()
        bundle.putString(AnalyticsKeys.PORTAL, portal)
        bundle.putString(AnalyticsKeys.ON_DEVICE, "false")
        bundle.putString(AnalyticsKeys.FILE_EXT, extension)
        addAnalytics(AnalyticsEvents.OPEN_EDITOR, bundle)
    }

    @JvmStatic
    fun addAnalyticsOpenExternal(portal: String, extension: String) {
        val bundle = Bundle()
        bundle.putString(AnalyticsKeys.PORTAL, portal)
        bundle.putString(AnalyticsKeys.ON_DEVICE, "false")
        bundle.putString(AnalyticsKeys.FILE_EXT, extension)
        addAnalytics(AnalyticsEvents.OPEN_EXTERNAL, bundle)
    }

    object AnalyticsEvents {
        const val CREATE_PORTAL = "portal_create"
        const val CREATE_ENTITY = "create_entity"
        const val CHECK_PORTAL = "check_portal"
        const val LOGIN_PORTAL = "portal_login"
        const val SWITCH_ACCOUNT = "account_switch"
        const val OPEN_PDF = "open_pdf"
        const val OPEN_EDITOR = "open_editor"
        const val OPEN_MEDIA = "open_media"
        const val OPEN_EXTERNAL = "open_external"
        const val OPERATION_RESULT = "operation_result"
        const val OPERATION_DETAILS = "operation_details"
    }

    object AnalyticsKeys {
        const val NONE = "none"
        const val SUCCESS = "success"
        const val FAILED = "failed"
        const val PORTAL = "portal"
        const val LOGIN = "email"
        const val PROVIDER = "provider"
        const val ON_DEVICE = "onDevice"
        const val TYPE = "type"
        const val FILE_EXT = "fileExt"
        const val TYPE_FILE = "file"
        const val TYPE_FOLDER = "folder"
    }

    interface OnRatingApp {
        fun onRatingApp(isRating: Boolean)
    }
}